packages/ai/src/types.ts

```diff
@@ -17,61 +19,118 @@ export interface QueryContext {
   max_object_count?: number;
 }
 
-export interface RAGRequest {
-  model: string;
-  prompt?: Prompt;
-  context: QueryContext;
-  query: string;
+interface RagRequestBase {
   stream?: boolean;
+  [key: string]: unknown;
 }
 
-export interface MessageStart {
-  type: "message_start";
-  message: {
-    role: "assistant" | "system" | "user";
-    id: string;
-    model: string;
-  };
-}
+export type RagRequest =
+  | (RagRequestBase & { messages: EdgeDBMessage[]; prompt?: undefined })
+  | (RagRequestBase & { prompt: string; messages?: undefined });
```

Comment:
I think we want:

```suggestion
  | (RagRequestBase & { messages: EdgeDBMessage[] })
  | (RagRequestBase & { prompt: string });
```

Right?

Reply:
yes yes!

Reply:
Yeah, you'll have to do:

```ts
if ("messages" in request) {
```

etc.

I mostly think it's a code smell to explicitly assign something as `undefined`. Maybe a compromise that keeps the ergonomics here is to use `null` and make it non-optional so there is no ambiguity? You'll probably have to do some work to conform to that type though, right?

One other more principled solution is to make this a more explicit discriminant by adding a `type` or `tag` or `kind` property to make it easier to detect which logical branch we're in without having to do a bunch of defensive checks for undefined/null.

Reply:
fixed this, I actually also needed to use the function:
```
export function isPromptRequest(
  request: RagRequest,
): request is RagRequestPrompt {
  return "prompt" in request;
}
```
To explicitly say that the type is RagRequestPrompt if `"prompt" in request`

Using just the `in` check is not enough, I guess it is still not able to differentiate the type .

I don't really like using the additional property like `type` in this case since the `RagRequest` comes from the client, I can tho probably update it in the `queryRag/streamRag` and send the type too to the `fetchRag`.

Reply:
Yeah, using a custom type guard here makes sense to me, good thinking.